### PR TITLE
🏃[e2e] add ClusterProvider and ClusterProxy

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR introduces it is required to have a cleaner separation between:

- A ClusterProvider, with its implementation based on kind
- A ClusterProxy that can act as an intermediary with an existing Kubernetes cluster, no matter if the cluster was created by the aforementioned ClusterProvider, by Cluster API (a workload cluster or a self-hosted cluster) or else.

This change is required in order to implement tests on self-hosted clusters (more details in the issue)

**Which issue(s) this PR fixes**:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/2910

xref #2753
xref #2637
xref #2636

/assign @vincepri
/assign @sedefsavas

### CODE REVIEW

File: test/framework/clusterctl/management_cluster.go [hunk 1]
```diff
@@ -1,194 +0,0 @@
...
-
-	// Scheme is used to initialize the scheme for the management cluster client.
-	Scheme *runtime.Scheme
-
-	// NewManagementClusterFn should return a new management cluster.
-	NewManagementClusterFn func(name string, scheme *runtime.Scheme) (cluster framework.ManagementCluster, kubeConfigPath string, err error)
-}
-
-// InitManagementCluster returns a new cluster initialized and the path to the kubeConfig file to be used to access it.
-func InitManagementCluster(ctx context.Context, input *InitManagementClusterInput) (framework.ManagementCluster, string) {
```
Comment (detiber) [inline test/framework/clusterctl/management_cluster.go:65]: It looks like this is currently being used by at least cluster-api-provider-vsphere.

Comment (fabriziopandini) [reply test/framework/clusterctl/management_cluster.go:65]: @detiber this is `test/framework/clusterctl.InitManagementCluster`, while in CAPV they are using `test/framework.InitManagementCluster` (without clusterctl)

Comment (detiber) [reply test/framework/clusterctl/management_cluster.go:65]: ah, carry on then :)

File: test/framework/bootstrap/kind_provider.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+}
+
+// Create a Kubernetes cluster using kind.
+func (k *kindClusterProvider) Create(ctx context.Context) {
+	Expect(ctx).NotTo(BeNil(), "ctx is required for Create")
+
+	// Sets the kubeconfig path to a temp file.
+	// NB. the ClusterProvider is responsible for the cleanup of this file
+	f, err := ioutil.TempFile("", "e2e-kind")
+	Expect(err).ToNot(HaveOccurred(), "Failed to create kubeconfig file for the kind cluster %q")
```
Comment (sedefsavas) [inline test/framework/bootstrap/kind_provider.go:78]: Missing cluster name here?

Comment (fabriziopandini) [reply test/framework/bootstrap/kind_provider.go]: Fixed!

File: test/framework/cluster_proxy.go [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+// by Cluster API (a workload cluster or a self-hosted cluster) or else.
+type ClusterProxy interface {
+	// GetName returns the name of the cluster.
+	GetName() string
+
+	// GetKubeconfigPath returns the path to the kubeconfig file to be used to access the Kubernetes cluster.
+	GetKubeconfigPath() string
+
+	// GetScheme returns the scheme defining the types hosted in the Kubernetes cluster.
+	// It is used when creating a
```
Comment (sedefsavas) [inline test/framework/cluster_proxy.go:53]: Unfinished comment.

File: test/framework/cluster_proxy.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+	GetWorkloadCluster(ctx context.Context, namespace, name string) ClusterProxy
+
+	// Dispose proxy's internal resources (the operation does not affects the Kubernetes cluster).
+	// This should be implemented as a synchronous function.
+	// Generally to be used in the AfterSuite function if a management cluster is shared between tests.
+	// Should try to clean everything up and report any dangling artifacts that needs manual intervention.
+	Dispose(context.Context)
+}
+
+// clusterProxy provides a base implementation of the ManagementCluster interface.
```
Comment (sedefsavas) [inline test/framework/cluster_proxy.go:75]:
```suggestion
// clusterProxy provides a base implementation of the ClusterProxy interface.
```

File: test/framework/cluster_proxy.go [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+func (p *clusterProxy) GetKubeconfigPath() string {
+	return p.kubeconfigPath
+}
+
+// GetScheme returns the scheme defining the types hosted in the cluster.
+func (p *clusterProxy) GetScheme() *runtime.Scheme {
+	return p.scheme
+}
+
+// GetClient returns a controller-runtime client for the management cluster.
```
Comment (sedefsavas) [inline test/framework/cluster_proxy.go:132]:
```suggestion
// GetClient returns a controller-runtime client for the cluster.
```

File: test/framework/cluster_proxy.go [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+func (p *clusterProxy) GetClient() client.Client {
+	config := p.getConfig()
+
+	c, err := client.New(config, client.Options{Scheme: p.scheme})
+	Expect(err).ToNot(HaveOccurred(), "Failed to get controller-runtime client")
+
+	return c
+}
+
+// GetClientSet returns a client-go client for the management cluster.
```
Comment (sedefsavas) [inline test/framework/cluster_proxy.go:142]:
```suggestion
// GetClientSet returns a client-go client for the cluster.
```

File: test/framework/cluster_proxy.go [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+
+	// Apply to apply YAML to the Kubernetes cluster, `kubectl apply`.
+	Apply(context.Context, []byte) error
+
+	// GetWorkloadCluster returns a proxy to a workload cluster defined in the Kubernetes cluster.
+	GetWorkloadCluster(ctx context.Context, namespace, name string) ClusterProxy
+
+	// Dispose proxy's internal resources (the operation does not affects the Kubernetes cluster).
+	// This should be implemented as a synchronous function.
+	// Generally to be used in the AfterSuite function if a management cluster is shared between tests.
```
Comment (sedefsavas) [inline test/framework/cluster_proxy.go:70]: Is this comment also relevant here?

Comment (fabriziopandini) [reply test/framework/cluster_proxy.go]: good catch, this was a leftover

**Comment by fabriziopandini** [comment]
@sedefsavas thanks for the feedback!
Comment addressed

**Pull request merged**